Improve level 4 to level 6 (#51)
abtion/guidelines: career/ladder.md

- [ ] Drives tasks from client research to client acceptance.

#### Simple
- [ ] Has a long track of delivering value for the client, finding the right balance between maintenance work and delivering features. They are recognized as someone who can solve a bottleneck when needed or reduce technical debt.
- [ ] Consistently demonstrates an understanding of the tradeoffs between technical, analytical, and product needs and leads to solutions that take all of these needs into account.

#### Test-driven

> - [ ] Performs TDD over Test Last (TLD) nearly all the time.

Reviewer: I'm not sure if this belongs this far up the ladder. 
Could this be on level 3 instead of `Moves from TFD to TFD+Refactor = TDD` 
https://github.com/abtion/guidelines/pull/51/files#diff-87ed49f2dc81712c4cd2c06d164f6f86dfe449980f817da28c9b559dbf1c9612R102

- [ ] Makes use of stubbing and mocking/understands tradeoffs of hitting the DB.
- [ ] Values and can speak to examples of high cohesion, low coupling.
- [ ] Frequently works with other engineers to reinforce a culture of testing and TDD.
- [ ] Identifies tests that are hard to maintain and fixes them proactively.
- [ ] Writes performance tests

#### Iterative
- [ ] Well-composed commit narratives in commit flows.

#### Communication
- [ ] PMs recognize them as someone who keeps the Asana board relevant. Internally, they act as PMs when needed.
- [ ] Have taught several developers to pair program.

#### Respectful teammate
- [ ] Makes others better through code reviews, thorough documentation, technical guidance, and mentoring.
- [ ] Communicate with PMs when goals are unattainable.
- [ ] Understands the difficulties that the team is going through and helps how they can.
- [ ] Respects the right of junior team members to own parts of the code.

#### Technical skills
- [ ] Demonstrates the ability to succeed in a wide range of complex situations across multiple axes: e.g., scale, uncertainty, interconnectedness.
- [ ] Researches and proposes new technologies to the team(s).

> - [ ] Go to expert in one area of technology: fx browser rendering or TDD.

Reviewer: ```suggestion
- [ ] Is known as a go-to expert in one area of technology: e.g. browser rendering or TDD.
```

- [ ] Drives tasks from client research to client acceptance.

#### Simple

> - [ ] Has a long track of delivering value for the client, finding the right balance between maintenance work and delivering features. They are recognized as someone who can solve a bottleneck when needed or reduce technical debt.

Reviewer 2: ```suggestion
- [ ] Has a long track of delivering value to the client, finding the right balance between maintenance work and delivering features. They are recognized as someone who can solve a bottleneck when needed or reduce technical debt.
```

#### Simple
- [ ] Creates architecture that enables many potential futures without knowing exactly what the future is.
- [ ] Understands the broad architecture of the entire system and is capable of explaining the technical decisions taken by the team.

#### Test-driven
- [ ] Writes tests in multiple languages
- [ ] Can speak from experience on how TDD influences class design.
- [ ] Able to identify and be comfortable deleting tests that no longer add value

> - [ ] Thinks about [F.I.R.S.T](https://github.com/ghsukumar/SFDC_Best_Practices/wiki/F.I.R.S.T-Principles-of-Unit-Testing) tests

Reviewer 3: Dead link. Might use this instead: https://tasdikrahman.me/2019/03/13/f-i-r-s-t-principles-of-testing/

- [ ] Does not take "all" responsibilities. They let their teammates take the responsibilities they care about.

#### Technical skills

> - [ ] Have turned around highly complex projects so that less experienced developers could collaborate on them. They are recognized for someone who knows how to transform a legacy system into a successful project.

Reviewer 2: ```suggestion
- [ ] Has turned around highly complex projects so that less experienced developers could collaborate on them. They are recognized for someone who knows how to transform a legacy system into a successful project.
```

- [ ] Focused on and responsible for their team's productivity and growth.

#### Simple
- [ ] Has solved bottlenecks in several projects throughout the years, making themselves replaceable after a planned period of time.
- [ ] Has mentored several teams on how to reduce the complexity of projects, services, and processes to get more done with less work.
- [ ] Excels at getting teams to focus on delivering high-impact value on the projects they mentor.

#### Test-driven
- [ ] Has succesfully laid out the test strategy on legacy projects. Has changed systems from unreliable to reliable. 
- [ ] Answers philosophical/economic questions like: was that test worth it for the business? Should e2e tests be run in multiple browsers if we support multiple browsers with the product? Why or why not?
- [ ] Facilitates dialog around reduced costs of implementation or maintenance with a team.
- [ ] Supports PMs in explaining the long term benefits of testing on client projects.
- [ ] Implements security tests.

#### Iterative
- [ ] Has brought several new working-methods that dramatically improve the overall company performance.

#### Communication

> - [ ] Collaborate with sales about high-level estimates and prioritization.

Reviewer: ```suggestion
- [ ] Collaborates with sales about high-level estimates and prioritization.
```

> - [ ] Has a long track of understanding the root cause of complex situations and building a roadmap that the team can follow.

Reviewer 3: What do you mean by the `building a roadmap` part.

Author: This focuses highly on team efficiency, rather than on the developer itself. When a project is complex, we want a principal engineer to be able to understand what it would take to change it and plan with the PM for the necessary change. The principal engineer will not necessarily be the developer who will end up doing the changes. Instead, other less-experienced developers will be able to solve the problem.

Reviewer 3: So something like `Has a long track record of suggesting ways to simplify complex apps, that the team can then implement`?

Author: Good suggestion. I'll change it.

#### Respectful teammate
- [ ] Mentors other engineers to reach their level (and pass them).

#### Technical skills

> - [ ] Recognized for high quality, impactful technical contributions.

Reviewer: ```suggestion
- [ ] Is recognized for high-quality, impactful technical contributions.
```
